tls/s2n_server_hello_retry.c

```diff
@@ -16,27 +16,77 @@
 
 #include "error/s2n_errno.h"
 #include "utils/s2n_blob.h"
+#include "tls/extensions/s2n_cookie.h"
+#include "tls/s2n_cipher_suites.h"
 #include "tls/s2n_tls.h"
+#include "tls/s2n_tls13_handshake.h"
 #include "utils/s2n_safety.h"
 
+/* From RFC5246 7.4.1.2. */
+#define S2N_TLS_COMPRESSION_METHOD_NULL 0
+
 /* from RFC: https://tools.ietf.org/html/rfc8446#section-4.1.3*/
-const uint8_t hello_retry_req_random[] = {
+static const uint8_t hello_retry_req_random[S2N_TLS_RANDOM_DATA_LEN] = {
     0xCF, 0x21, 0xAD, 0x74, 0xE5, 0x9A, 0x61, 0x11, 0xBE, 0x1D, 0x8C, 0x02, 0x1E, 0x65, 0xB8, 0x91,
     0xC2, 0xA2, 0x11, 0x16, 0x7A, 0xBB, 0x8C, 0x5E, 0x07, 0x9E, 0x09, 0xE2, 0xC8, 0xA8, 0x33, 0x9C
 };
 
-inline bool s2n_is_hello_retry_req(struct s2n_connection *conn)
+static int s2n_conn_reset_retry_values(struct s2n_connection *conn)
 {
-    return s2n_constant_time_equals(hello_retry_req_random, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN);
+    /* Reset handshake values */
+    conn->handshake.hello_retry_request = 1;
+    conn->handshake.client_hello_received = 0;
+
+    /* Reset client hello state */
+    GUARD(s2n_stuffer_wipe(&conn->client_hello.raw_message));
+    GUARD(s2n_stuffer_resize(&conn->client_hello.raw_message, 0));
+    GUARD(s2n_client_hello_free(&conn->client_hello));
+    GUARD(s2n_stuffer_growable_alloc(&conn->client_hello.raw_message, 0));
+
+    return 0;
 }
 
 
+/* Lets the client determine whether a HelloRetryRequest is valid */
+bool s2n_server_hello_retry_is_valid(struct s2n_connection *conn)
+{
+    bool has_versions_ext = conn->server_protocol_version > 0;
+    bool has_correct_random = (memcmp(hello_retry_req_random, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN) == 0);
+
+    return has_versions_ext && has_correct_random && conn->client_protocol_version == S2N_TLS13;
+}
+
 int s2n_server_hello_retry_send(struct s2n_connection *conn)
 {
-    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+    struct s2n_stuffer *out = &conn->handshake.io;
+    uint8_t protocol_version[S2N_TLS_PROTOCOL_VERSION_LEN];
+
+    /* We only send retries in S2N_TLS13, so we know the legacy protocol version should be S2N_TLS12 */
+    protocol_version[0] = (uint8_t)(S2N_TLS12 / 10);
+    protocol_version[1] = (uint8_t)(S2N_TLS12 % 10);
+    GUARD(s2n_stuffer_write_bytes(out, protocol_version, S2N_TLS_PROTOCOL_VERSION_LEN));
```

Comment:
this pattern is repeated a lot - a good chance for a refactor - a macro would do:

```c
S2N_WRITE_PROTOCOL_VERSION(out, version)  do { \
  uint8_t __protocol_version[S2N_TLS_PROTOCOL_VERSION_LEN] = { (uint8_t)( (version) / 10), (uint8_t)(version % 10) }; \
  GUARD(s2n_stuffer_write_bytes((out), __protocol_version, S2N_TLS_PROTOCOL_VERSION_LEN)); \
} while(0)
```

Then invoked as:

```c
S2N_WRITE_PROTOCOL_VERSION(out, S2N_TLS12);
```